compiler/circle-opselector/pass/OpSelector.cpp

```diff
@@ -0,0 +1,503 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "OpSelector.h"
+// #include "luci/../../src/CircleImportMetadata.h"
+// #include "luci/../../src/PostImport.h"
+
+#include <luci/Service/ChangeOutputs.h>
+
+#include <luci/Import/GraphBuilder.h>
+#include <luci/Import/GraphBuilderContext.h>
+#include <luci/Import/GraphBuilderRegistry.h>
+#include <luci/Import/CircleReader.h>
+#include <luci/Import/Nodes/CircleConst.h>
+
+#include <luci/Profile/CircleNodeID.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+#include <loco/IR/Graph.h>
+#include <luci/IR/CircleNode.h>
+#include <loco/IR/Node.h>
+
+#include <oops/UserExn.h>
+
+#include <cassert>
+#include <iostream>
+#include <map>
+
+namespace opselector
+{
+
+void convert_graph(const luci::GraphBuilderSource &source, luci::CircleReader &reader,
+                   loco::Graph *graph)
+{
+  auto nodefinder = std::make_unique<luci::IndexNodeFinder>();
+  auto tensoroutputs = std::make_unique<luci::IndexTensorOutputs>();
+
+  luci::GraphBuilderContext gb_context(graph, &reader, nodefinder.get(), tensoroutputs.get());
+
+  const auto &operators = reader.operators();
+  const auto &tensors = reader.tensors();
+  auto tensors_ptr = reader.tensors_ptr();
+  assert(tensors_ptr != nullptr);
+
+  // build a cache to identify if a tensor is output of an operator
+  // if this is set, we should not create a CircleConst for this tensor
+  for (uint32_t i = 0; i < operators.size(); ++i)
+  {
+    const circle::OperatorT &op = *operators[i];
+    const auto &outputs = op.outputs;
+
+    for (uint32_t j = 0; j < outputs.size(); ++j)
+    {
+      auto tidx = outputs[j];
+      tensoroutputs->enroll(tidx);
+    }
+  }
+
+  // graph inputs; there are no input nodes in TFlite but just Tensors
```

Comment:
I don't understand the sentence. What do you mean by _"there are no input nodes in TFlite but just Tensors"_? Each model has at least one `SubGraph` and each `SubGraph` has inputs and outputs.